// react-easy-crop/index.js
Object.defineProperties(exports, {
	__esModule: { value: true },
	[Symbol.toStringTag]: { value: "Module" }
});
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let react = require("react");
react = __toESM(react);
let normalize_wheel = require("normalize-wheel");
normalize_wheel = __toESM(normalize_wheel);
//#region \0@oxc-project+runtime@0.134.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}
//#endregion
//#region \0@oxc-project+runtime@0.134.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}
//#endregion
//#region \0@oxc-project+runtime@0.134.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}
//#endregion
//#region \0@oxc-project+runtime@0.134.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}
//#endregion
//#region \0@oxc-project+runtime@0.134.0/helpers/esm/objectSpread2.js
function ownKeys(e, r) {
	var t = Object.keys(e);
	if (Object.getOwnPropertySymbols) {
		var o = Object.getOwnPropertySymbols(e);
		r && (o = o.filter(function(r) {
			return Object.getOwnPropertyDescriptor(e, r).enumerable;
		})), t.push.apply(t, o);
	}
	return t;
}
function _objectSpread2(e) {
	for (var r = 1; r < arguments.length; r++) {
		var t = null != arguments[r] ? arguments[r] : {};
		r % 2 ? ownKeys(Object(t), !0).forEach(function(r) {
			_defineProperty(e, r, t[r]);
		}) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function(r) {
			Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
		});
	}
	return e;
}
//#endregion
//#region src/helpers.ts
/**
* Compute the dimension of the crop area based on media size,
* aspect ratio and optionally rotation
*/
function getCropSize(mediaWidth, mediaHeight, containerWidth, containerHeight, aspect, rotation = 0) {
	const { width, height } = rotateSize(mediaWidth, mediaHeight, rotation);
	const fittingWidth = Math.min(width, containerWidth);
	const fittingHeight = Math.min(height, containerHeight);
	if (fittingWidth > fittingHeight * aspect) return {
		width: fittingHeight * aspect,
		height: fittingHeight
	};
	return {
		width: fittingWidth,
		height: fittingWidth / aspect
	};
}
/**
* Compute media zoom.
* We fit the media into the container with "max-width: 100%; max-height: 100%;"
*/
function getMediaZoom(mediaSize) {
	return mediaSize.width > mediaSize.height ? mediaSize.width / mediaSize.naturalWidth : mediaSize.height / mediaSize.naturalHeight;
}
/**
* Ensure a new media position stays in the crop area.
*/
function restrictPosition(position, mediaSize, cropSize, zoom, rotation = 0) {
	const { width, height } = rotateSize(mediaSize.width, mediaSize.height, rotation);
	return {
		x: restrictPositionCoord(position.x, width, cropSize.width, zoom),
		y: restrictPositionCoord(position.y, height, cropSize.height, zoom)
	};
}
function restrictPositionCoord(position, mediaSize, cropSize, zoom) {
	const maxPosition = Math.abs(mediaSize * zoom / 2 - cropSize / 2);
	return clamp(position, -maxPosition, maxPosition);
}
function getDistanceBetweenPoints(pointA, pointB) {
	return Math.sqrt(Math.pow(pointA.y - pointB.y, 2) + Math.pow(pointA.x - pointB.x, 2));
}
function getRotationBetweenPoints(pointA, pointB) {
	return Math.atan2(pointB.y - pointA.y, pointB.x - pointA.x) * 180 / Math.PI;
}
/**
* Compute the output cropped area of the media in percentages and pixels.
* x/y are the top-left coordinates on the src media
*/
function computeCroppedArea(crop, mediaSize, cropSize, aspect, zoom, rotation = 0, restrictPosition = true) {
	const limitAreaFn = restrictPosition ? limitArea : noOp;
	const mediaBBoxSize = rotateSize(mediaSize.width, mediaSize.height, rotation);
	const mediaNaturalBBoxSize = rotateSize(mediaSize.naturalWidth, mediaSize.naturalHeight, rotation);
	const croppedAreaPercentages = {
		x: limitAreaFn(100, ((mediaBBoxSize.width - cropSize.width / zoom) / 2 - crop.x / zoom) / mediaBBoxSize.width * 100),
		y: limitAreaFn(100, ((mediaBBoxSize.height - cropSize.height / zoom) / 2 - crop.y / zoom) / mediaBBoxSize.height * 100),
		width: limitAreaFn(100, cropSize.width / mediaBBoxSize.width * 100 / zoom),
		height: limitAreaFn(100, cropSize.height / mediaBBoxSize.height * 100 / zoom)
	};
	const widthInPixels = Math.round(limitAreaFn(mediaNaturalBBoxSize.width, croppedAreaPercentages.width * mediaNaturalBBoxSize.width / 100));
	const heightInPixels = Math.round(limitAreaFn(mediaNaturalBBoxSize.height, croppedAreaPercentages.height * mediaNaturalBBoxSize.height / 100));
	const sizePixels = mediaNaturalBBoxSize.width >= mediaNaturalBBoxSize.height * aspect ? {
		width: Math.round(heightInPixels * aspect),
		height: heightInPixels
	} : {
		width: widthInPixels,
		height: Math.round(widthInPixels / aspect)
	};
	return {
		croppedAreaPercentages,
		croppedAreaPixels: _objectSpread2(_objectSpread2({}, sizePixels), {}, {
			x: Math.round(limitAreaFn(mediaNaturalBBoxSize.width - sizePixels.width, croppedAreaPercentages.x * mediaNaturalBBoxSize.width / 100)),
			y: Math.round(limitAreaFn(mediaNaturalBBoxSize.height - sizePixels.height, croppedAreaPercentages.y * mediaNaturalBBoxSize.height / 100))
		})
	};
}
/**
* Ensure the returned value is between 0 and max
*/
function limitArea(max, value) {
	return Math.min(max, Math.max(0, value));
}
function noOp(_max, value) {
	return value;
}
/**
* Compute crop and zoom from the croppedAreaPercentages.
*/
function getInitialCropFromCroppedAreaPercentages(croppedAreaPercentages, mediaSize, rotation, cropSize, minZoom, maxZoom) {
	const mediaBBoxSize = rotateSize(mediaSize.width, mediaSize.height, rotation);
	const zoom = clamp(cropSize.width / mediaBBoxSize.width * (100 / croppedAreaPercentages.width), minZoom, maxZoom);
	return {
		crop: {
			x: zoom * mediaBBoxSize.width / 2 - cropSize.width / 2 - mediaBBoxSize.width * zoom * (croppedAreaPercentages.x / 100),
			y: zoom * mediaBBoxSize.height / 2 - cropSize.height / 2 - mediaBBoxSize.height * zoom * (croppedAreaPercentages.y / 100)
		},
		zoom
	};
}
/**
* Compute zoom from the croppedAreaPixels
*/
function getZoomFromCroppedAreaPixels(croppedAreaPixels, mediaSize, cropSize) {
	const mediaZoom = getMediaZoom(mediaSize);
	return cropSize.height > cropSize.width ? cropSize.height / (croppedAreaPixels.height * mediaZoom) : cropSize.width / (croppedAreaPixels.width * mediaZoom);
}
/**
* Compute crop and zoom from the croppedAreaPixels
*/
function getInitialCropFromCroppedAreaPixels(croppedAreaPixels, mediaSize, rotation = 0, cropSize, minZoom, maxZoom) {
	const mediaNaturalBBoxSize = rotateSize(mediaSize.naturalWidth, mediaSize.naturalHeight, rotation);
	const zoom = clamp(getZoomFromCroppedAreaPixels(croppedAreaPixels, mediaSize, cropSize), minZoom, maxZoom);
	const cropZoom = cropSize.height > cropSize.width ? cropSize.height / croppedAreaPixels.height : cropSize.width / croppedAreaPixels.width;
	return {
		crop: {
			x: ((mediaNaturalBBoxSize.width - croppedAreaPixels.width) / 2 - croppedAreaPixels.x) * cropZoom,
			y: ((mediaNaturalBBoxSize.height - croppedAreaPixels.height) / 2 - croppedAreaPixels.y) * cropZoom
		},
		zoom
	};
}
/**
* Return the point that is the center of point a and b
*/
function getCenter(a, b) {
	return {
		x: (b.x + a.x) / 2,
		y: (b.y + a.y) / 2
	};
}
function getRadianAngle(degreeValue) {
	return degreeValue * Math.PI / 180;
}
/**
* Returns the new bounding area of a rotated rectangle.
*/
function rotateSize(width, height, rotation) {
	const rotRad = getRadianAngle(rotation);
	return {
		width: Math.abs(Math.cos(rotRad) * width) + Math.abs(Math.sin(rotRad) * height),
		height: Math.abs(Math.sin(rotRad) * width) + Math.abs(Math.cos(rotRad) * height)
	};
}
/**
* Clamp value between min and max
*/
function clamp(value, min, max) {
	return Math.min(Math.max(value, min), max);
}
/**
* Combine multiple class names into a single string.
*/
function classNames(...args) {
	return args.filter((value) => {
		if (typeof value === "string" && value.length > 0) return true;
		return false;
	}).join(" ").trim();
}
//#endregion
//#region src/styles.css?raw
var styles_default = ".reactEasyCrop_Container {\n  position: absolute;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  overflow: hidden;\n  user-select: none;\n  touch-action: none;\n  cursor: move;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n}\n\n.reactEasyCrop_Image,\n.reactEasyCrop_Video {\n  will-change: transform; /* this improves performances and prevent painting issues on iOS Chrome */\n  max-width: unset; /* prevent global img/video reset rules from constraining the cropper media */\n}\n\n.reactEasyCrop_Contain {\n  max-width: 100%;\n  max-height: 100%;\n  margin: auto;\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  left: 0;\n  right: 0;\n}\n.reactEasyCrop_Cover_Horizontal {\n  width: 100%;\n  height: auto;\n}\n.reactEasyCrop_Cover_Vertical {\n  width: auto;\n  height: 100%;\n}\n\n.reactEasyCrop_CropArea {\n  position: absolute;\n  left: 50%;\n  top: 50%;\n  transform: translate(-50%, -50%);\n  border: 1px solid rgba(255, 255, 255, 0.5);\n  box-sizing: border-box;\n  box-shadow: 0 0 0 9999em;\n  color: rgba(0, 0, 0, 0.5);\n  overflow: hidden;\n}\n\n.reactEasyCrop_CropAreaRound {\n  border-radius: 50%;\n}\n\n.reactEasyCrop_CropAreaGrid::before {\n  content: ' ';\n  box-sizing: border-box;\n  position: absolute;\n  border: 1px solid rgba(255, 255, 255, 0.5);\n  top: 0;\n  bottom: 0;\n  left: 33.33%;\n  right: 33.33%;\n  border-top: 0;\n  border-bottom: 0;\n}\n\n.reactEasyCrop_CropAreaGrid::after {\n  content: ' ';\n  box-sizing: border-box;\n  position: absolute;\n  border: 1px solid rgba(255, 255, 255, 0.5);\n  top: 33.33%;\n  bottom: 33.33%;\n  left: 0;\n  right: 0;\n  border-left: 0;\n  border-right: 0;\n}\n";
//#endregion
//#region src/Cropper.tsx
const MIN_ZOOM = 1;
const MAX_ZOOM = 3;
const KEYBOARD_STEP = 1;
var Cropper = class Cropper extends react.Component {
	constructor(..._args) {
		super(..._args);
		this.cropperRef = react.createRef();
		this.imageRef = react.createRef();
		this.videoRef = react.createRef();
		this.containerPosition = {
			x: 0,
			y: 0
		};
		this.containerRef = null;
		this.styleRef = null;
		this.containerRect = null;
		this.mediaSize = {
			width: 0,
			height: 0,
			naturalWidth: 0,
			naturalHeight: 0
		};
		this.dragStartPosition = {
			x: 0,
			y: 0
		};
		this.dragStartCrop = {
			x: 0,
			y: 0
		};
		this.gestureZoomStart = 0;
		this.gestureRotationStart = 0;
		this.isTouching = false;
		this.lastPinchDistance = 0;
		this.lastPinchRotation = 0;
		this.rafDragTimeout = null;
		this.rafPinchTimeout = null;
		this.wheelTimer = null;
		this.currentDoc = typeof document !== "undefined" ? document : null;
		this.currentWindow = typeof window !== "undefined" ? window : null;
		this.resizeObserver = null;
		this.previousCropSize = null;
		this.isInitialized = false;
		this.state = {
			cropSize: null,
			hasWheelJustStarted: false,
			mediaObjectFit: void 0
		};
		this.initResizeObserver = () => {
			if (typeof window.ResizeObserver === "undefined" || !this.containerRef) return;
			let isFirstResize = true;
			this.resizeObserver = new window.ResizeObserver((entries) => {
				if (isFirstResize) {
					isFirstResize = false;
					return;
				}
				this.computeSizes();
			});
			this.resizeObserver.observe(this.containerRef);
		};
		this.preventZoomSafari = (e) => e.preventDefault();
		this.cleanEvents = () => {
			if (!this.currentDoc) return;
			this.currentDoc.removeEventListener("mousemove", this.onMouseMove);
			this.currentDoc.removeEventListener("mouseup", this.onDragStopped);
			this.currentDoc.removeEventListener("touchmove", this.onTouchMove);
			this.currentDoc.removeEventListener("touchend", this.onDragStopped);
			this.currentDoc.removeEventListener("gesturechange", this.onGestureChange);
			this.currentDoc.removeEventListener("gestureend", this.onGestureEnd);
			this.currentDoc.removeEventListener("scroll", this.onScroll);
		};
		this.clearScrollEvent = () => {
			if (this.containerRef) this.containerRef.removeEventListener("wheel", this.onWheel);
			if (this.wheelTimer) clearTimeout(this.wheelTimer);
		};
		this.onMediaLoad = () => {
			const cropSize = this.computeSizes();
			if (cropSize) {
				this.previousCropSize = cropSize;
				this.emitCropData();
				this.setInitialCrop(cropSize);
				this.isInitialized = true;
			}
			if (this.props.onMediaLoaded) this.props.onMediaLoaded(this.mediaSize);
		};
		this.setInitialCrop = (cropSize) => {
			if (this.props.initialCroppedAreaPercentages) {
				const { crop, zoom } = getInitialCropFromCroppedAreaPercentages(this.props.initialCroppedAreaPercentages, this.mediaSize, this.props.rotation, cropSize, this.props.minZoom, this.props.maxZoom);
				this.props.onCropChange(crop);
				this.props.onZoomChange && this.props.onZoomChange(zoom);
			} else if (this.props.initialCroppedAreaPixels) {
				const { crop, zoom } = getInitialCropFromCroppedAreaPixels(this.props.initialCroppedAreaPixels, this.mediaSize, this.props.rotation, cropSize, this.props.minZoom, this.props.maxZoom);
				this.props.onCropChange(crop);
				this.props.onZoomChange && this.props.onZoomChange(zoom);
			}
		};
		this.computeSizes = () => {
			const mediaRef = this.imageRef.current || this.videoRef.current;
			if (mediaRef && this.containerRef) {
				var _this$imageRef$curren, _this$videoRef$curren, _this$imageRef$curren2, _this$videoRef$curren2, _this$state$cropSize, _this$state$cropSize2;
				this.containerRect = this.containerRef.getBoundingClientRect();
				this.saveContainerPosition();
				const containerAspect = this.containerRect.width / this.containerRect.height;
				const naturalWidth = ((_this$imageRef$curren = this.imageRef.current) === null || _this$imageRef$curren === void 0 ? void 0 : _this$imageRef$curren.naturalWidth) || ((_this$videoRef$curren = this.videoRef.current) === null || _this$videoRef$curren === void 0 ? void 0 : _this$videoRef$curren.videoWidth) || 0;
				const naturalHeight = ((_this$imageRef$curren2 = this.imageRef.current) === null || _this$imageRef$curren2 === void 0 ? void 0 : _this$imageRef$curren2.naturalHeight) || ((_this$videoRef$curren2 = this.videoRef.current) === null || _this$videoRef$curren2 === void 0 ? void 0 : _this$videoRef$curren2.videoHeight) || 0;
				const isMediaScaledDown = mediaRef.offsetWidth < naturalWidth || mediaRef.offsetHeight < naturalHeight;
				const mediaAspect = naturalWidth / naturalHeight;
				let renderedMediaSize;
				if (isMediaScaledDown) switch (this.state.mediaObjectFit) {
					default:
					case "contain":
						renderedMediaSize = containerAspect > mediaAspect ? {
							width: this.containerRect.height * mediaAspect,
							height: this.containerRect.height
						} : {
							width: this.containerRect.width,
							height: this.containerRect.width / mediaAspect
						};
						break;
					case "horizontal-cover":
						renderedMediaSize = {
							width: this.containerRect.width,
							height: this.containerRect.width / mediaAspect
						};
						break;
					case "vertical-cover":
						renderedMediaSize = {
							width: this.containerRect.height * mediaAspect,
							height: this.containerRect.height
						};
						break;
				}
				else renderedMediaSize = {
					width: mediaRef.offsetWidth,
					height: mediaRef.offsetHeight
				};
				this.mediaSize = _objectSpread2(_objectSpread2({}, renderedMediaSize), {}, {
					naturalWidth,
					naturalHeight
				});
				if (this.props.setMediaSize) this.props.setMediaSize(this.mediaSize);
				const cropSize = this.props.cropSize ? this.props.cropSize : getCropSize(this.mediaSize.width, this.mediaSize.height, this.containerRect.width, this.containerRect.height, this.props.aspect, this.props.rotation);
				if (((_this$state$cropSize = this.state.cropSize) === null || _this$state$cropSize === void 0 ? void 0 : _this$state$cropSize.height) !== cropSize.height || ((_this$state$cropSize2 = this.state.cropSize) === null || _this$state$cropSize2 === void 0 ? void 0 : _this$state$cropSize2.width) !== cropSize.width) this.props.onCropSizeChange && this.props.onCropSizeChange(cropSize);
				this.setState({ cropSize }, this.recomputeCropPosition);
				if (this.props.setCropSize) this.props.setCropSize(cropSize);
				return cropSize;
			}
		};
		this.saveContainerPosition = () => {
			if (this.containerRef) {
				const bounds = this.containerRef.getBoundingClientRect();
				this.containerPosition = {
					x: bounds.left,
					y: bounds.top
				};
			}
		};
		this.onMouseDown = (e) => {
			if (!this.currentDoc) return;
			e.preventDefault();
			this.currentDoc.addEventListener("mousemove", this.onMouseMove);
			this.currentDoc.addEventListener("mouseup", this.onDragStopped);
			this.saveContainerPosition();
			this.onDragStart(Cropper.getMousePoint(e));
		};
		this.onMouseMove = (e) => this.onDrag(Cropper.getMousePoint(e));
		this.onScroll = (e) => {
			if (!this.currentDoc) return;
			e.preventDefault();
			this.saveContainerPosition();
		};
		this.onTouchStart = (e) => {
			if (!this.currentDoc) return;
			this.isTouching = true;
			if (this.props.onTouchRequest && !this.props.onTouchRequest(e)) return;
			this.currentDoc.addEventListener("touchmove", this.onTouchMove, { passive: false });
			this.currentDoc.addEventListener("touchend", this.onDragStopped);
			this.saveContainerPosition();
			if (e.touches.length === 2) this.onPinchStart(e);
			else if (e.touches.length === 1) this.onDragStart(Cropper.getTouchPoint(e.touches[0]));
		};
		this.onTouchMove = (e) => {
			e.preventDefault();
			if (e.touches.length === 2) this.onPinchMove(e);
			else if (e.touches.length === 1) this.onDrag(Cropper.getTouchPoint(e.touches[0]));
		};
		this.onGestureStart = (e) => {
			if (!this.currentDoc) return;
			e.preventDefault();
			this.currentDoc.addEventListener("gesturechange", this.onGestureChange);
			this.currentDoc.addEventListener("gestureend", this.onGestureEnd);
			this.gestureZoomStart = this.props.zoom;
			this.gestureRotationStart = this.props.rotation;
		};
		this.onGestureChange = (e) => {
			e.preventDefault();
			if (this.isTouching) return;
			const point = Cropper.getMousePoint(e);
			const newZoom = this.gestureZoomStart - 1 + e.scale;
			this.setNewZoom(newZoom, point, { shouldUpdatePosition: true });
			if (this.props.onRotationChange) {
				const newRotation = this.gestureRotationStart + e.rotation;
				this.props.onRotationChange(newRotation);
			}
		};
		this.onGestureEnd = (e) => {
			this.cleanEvents();
		};
		this.onDragStart = ({ x, y }) => {
			var _this$props$onInterac, _this$props;
			this.dragStartPosition = {
				x,
				y
			};
			this.dragStartCrop = _objectSpread2({}, this.props.crop);
			(_this$props$onInterac = (_this$props = this.props).onInteractionStart) === null || _this$props$onInterac === void 0 || _this$props$onInterac.call(_this$props);
		};
		this.onDrag = ({ x, y }) => {
			if (!this.currentWindow) return;
			if (this.rafDragTimeout) this.currentWindow.cancelAnimationFrame(this.rafDragTimeout);
			this.rafDragTimeout = this.currentWindow.requestAnimationFrame(() => {
				if (!this.state.cropSize) return;
				if (x === void 0 || y === void 0) return;
				const offsetX = x - this.dragStartPosition.x;
				const offsetY = y - this.dragStartPosition.y;
				const requestedPosition = {
					x: this.dragStartCrop.x + offsetX,
					y: this.dragStartCrop.y + offsetY
				};
				const newPosition = this.props.restrictPosition ? restrictPosition(requestedPosition, this.mediaSize, this.state.cropSize, this.props.zoom, this.props.rotation) : requestedPosition;
				this.props.onCropChange(newPosition);
			});
		};
		this.onDragStopped = () => {
			var _this$props$onInterac2, _this$props2;
			this.isTouching = false;
			this.cleanEvents();
			this.emitCropData();
			(_this$props$onInterac2 = (_this$props2 = this.props).onInteractionEnd) === null || _this$props$onInterac2 === void 0 || _this$props$onInterac2.call(_this$props2);
		};
		this.onWheel = (e) => {
			if (!this.currentWindow) return;
			if (this.props.onWheelRequest && !this.props.onWheelRequest(e)) return;
			e.preventDefault();
			const point = Cropper.getMousePoint(e);
			const { pixelY } = (0, normalize_wheel.default)(e);
			const newZoom = this.props.zoom - pixelY * this.props.zoomSpeed / 200;
			this.setNewZoom(newZoom, point, { shouldUpdatePosition: true });
			if (!this.state.hasWheelJustStarted) this.setState({ hasWheelJustStarted: true }, () => {
				var _this$props$onInterac3, _this$props3;
				return (_this$props$onInterac3 = (_this$props3 = this.props).onInteractionStart) === null || _this$props$onInterac3 === void 0 ? void 0 : _this$props$onInterac3.call(_this$props3);
			});
			if (this.wheelTimer) clearTimeout(this.wheelTimer);
			this.wheelTimer = this.currentWindow.setTimeout(() => this.setState({ hasWheelJustStarted: false }, () => {
				var _this$props$onInterac4, _this$props4;
				return (_this$props$onInterac4 = (_this$props4 = this.props).onInteractionEnd) === null || _this$props$onInterac4 === void 0 ? void 0 : _this$props$onInterac4.call(_this$props4);
			}), 250);
		};
		this.getPointOnContainer = ({ x, y }, containerTopLeft) => {
			if (!this.containerRect) throw new Error("The Cropper is not mounted");
			return {
				x: this.containerRect.width / 2 - (x - containerTopLeft.x),
				y: this.containerRect.height / 2 - (y - containerTopLeft.y)
			};
		};
		this.getPointOnMedia = ({ x, y }) => {
			const { crop, zoom } = this.props;
			return {
				x: (x + crop.x) / zoom,
				y: (y + crop.y) / zoom
			};
		};
		this.setNewZoom = (zoom, point, { shouldUpdatePosition = true } = {}) => {
			if (!this.state.cropSize || !this.props.onZoomChange) return;
			const newZoom = clamp(zoom, this.props.minZoom, this.props.maxZoom);
			if (shouldUpdatePosition) {
				const zoomPoint = this.getPointOnContainer(point, this.containerPosition);
				const zoomTarget = this.getPointOnMedia(zoomPoint);
				const requestedPosition = {
					x: zoomTarget.x * newZoom - zoomPoint.x,
					y: zoomTarget.y * newZoom - zoomPoint.y
				};
				const newPosition = this.props.restrictPosition ? restrictPosition(requestedPosition, this.mediaSize, this.state.cropSize, newZoom, this.props.rotation) : requestedPosition;
				this.props.onCropChange(newPosition);
			}
			this.props.onZoomChange(newZoom);
		};
		this.getCropData = () => {
			if (!this.state.cropSize) return null;
			return computeCroppedArea(this.props.restrictPosition ? restrictPosition(this.props.crop, this.mediaSize, this.state.cropSize, this.props.zoom, this.props.rotation) : this.props.crop, this.mediaSize, this.state.cropSize, this.getAspect(), this.props.zoom, this.props.rotation, this.props.restrictPosition);
		};
		this.emitCropData = () => {
			const cropData = this.getCropData();
			if (!cropData) return;
			const { croppedAreaPercentages, croppedAreaPixels } = cropData;
			if (this.props.onCropComplete) this.props.onCropComplete(croppedAreaPercentages, croppedAreaPixels);
			if (this.props.onCropAreaChange) this.props.onCropAreaChange(croppedAreaPercentages, croppedAreaPixels);
		};
		this.emitCropAreaChange = () => {
			const cropData = this.getCropData();
			if (!cropData) return;
			const { croppedAreaPercentages, croppedAreaPixels } = cropData;
			if (this.props.onCropAreaChange) this.props.onCropAreaChange(croppedAreaPercentages, croppedAreaPixels);
		};
		this.recomputeCropPosition = () => {
			var _this$previousCropSiz, _this$previousCropSiz2;
			if (!this.state.cropSize) return;
			let adjustedCrop = this.props.crop;
			if (this.isInitialized && ((_this$previousCropSiz = this.previousCropSize) === null || _this$previousCropSiz === void 0 ? void 0 : _this$previousCropSiz.width) && ((_this$previousCropSiz2 = this.previousCropSize) === null || _this$previousCropSiz2 === void 0 ? void 0 : _this$previousCropSiz2.height)) {
				if (Math.abs(this.previousCropSize.width - this.state.cropSize.width) > 1e-6 || Math.abs(this.previousCropSize.height - this.state.cropSize.height) > 1e-6) {
					const scaleX = this.state.cropSize.width / this.previousCropSize.width;
					const scaleY = this.state.cropSize.height / this.previousCropSize.height;
					adjustedCrop = {
						x: this.props.crop.x * scaleX,
						y: this.props.crop.y * scaleY
					};
				}
			}
			const newPosition = this.props.restrictPosition ? restrictPosition(adjustedCrop, this.mediaSize, this.state.cropSize, this.props.zoom, this.props.rotation) : adjustedCrop;
			this.previousCropSize = this.state.cropSize;
			this.props.onCropChange(newPosition);
			this.emitCropData();
		};
		this.onKeyDown = (event) => {
			const { crop, onCropChange, keyboardStep, zoom, rotation } = this.props;
			let step = keyboardStep;
			if (!this.state.cropSize) return;
			if (event.shiftKey) step *= .2;
			let newCrop = _objectSpread2({}, crop);
			switch (event.key) {
				case "ArrowUp":
					newCrop.y -= step;
					event.preventDefault();
					break;
				case "ArrowDown":
					newCrop.y += step;
					event.preventDefault();
					break;
				case "ArrowLeft":
					newCrop.x -= step;
					event.preventDefault();
					break;
				case "ArrowRight":
					newCrop.x += step;
					event.preventDefault();
					break;
				default: return;
			}
			if (this.props.restrictPosition) newCrop = restrictPosition(newCrop, this.mediaSize, this.state.cropSize, zoom, rotation);
			if (!event.repeat) {
				var _this$props$onInterac5, _this$props5;
				(_this$props$onInterac5 = (_this$props5 = this.props).onInteractionStart) === null || _this$props$onInterac5 === void 0 || _this$props$onInterac5.call(_this$props5);
			}
			onCropChange(newCrop);
		};
		this.onKeyUp = (event) => {
			var _this$props$onInterac6, _this$props6;
			switch (event.key) {
				case "ArrowUp":
				case "ArrowDown":
				case "ArrowLeft":
				case "ArrowRight":
					event.preventDefault();
					break;
				default: return;
			}
			this.emitCropData();
			(_this$props$onInterac6 = (_this$props6 = this.props).onInteractionEnd) === null || _this$props$onInterac6 === void 0 || _this$props$onInterac6.call(_this$props6);
		};
	}
	componentDidMount() {
		if (!this.currentDoc || !this.currentWindow) return;
		if (this.containerRef) {
			if (this.containerRef.ownerDocument) this.currentDoc = this.containerRef.ownerDocument;
			if (this.currentDoc.defaultView) this.currentWindow = this.currentDoc.defaultView;
			this.initResizeObserver();
			if (typeof window.ResizeObserver === "undefined") this.currentWindow.addEventListener("resize", this.computeSizes);
			this.props.zoomWithScroll && this.containerRef.addEventListener("wheel", this.onWheel, { passive: false });
			this.containerRef.addEventListener("gesturestart", this.onGestureStart);
		}
		this.currentDoc.addEventListener("scroll", this.onScroll);
		if (!this.props.disableAutomaticStylesInjection) {
			this.styleRef = this.currentDoc.createElement("style");
			this.styleRef.setAttribute("type", "text/css");
			if (this.props.nonce) this.styleRef.setAttribute("nonce", this.props.nonce);
			this.styleRef.innerHTML = styles_default;
			this.currentDoc.head.appendChild(this.styleRef);
		}
		if (this.imageRef.current && this.imageRef.current.complete) this.onMediaLoad();
		if (this.props.setImageRef) this.props.setImageRef(this.imageRef);
		if (this.props.setVideoRef) this.props.setVideoRef(this.videoRef);
		if (this.props.setCropperRef) this.props.setCropperRef(this.cropperRef);
	}
	componentWillUnmount() {
		var _this$resizeObserver;
		if (!this.currentDoc || !this.currentWindow) return;
		if (typeof window.ResizeObserver === "undefined") this.currentWindow.removeEventListener("resize", this.computeSizes);
		(_this$resizeObserver = this.resizeObserver) === null || _this$resizeObserver === void 0 || _this$resizeObserver.disconnect();
		if (this.containerRef) this.containerRef.removeEventListener("gesturestart", this.preventZoomSafari);
		if (this.styleRef) {
			var _this$styleRef$parent;
			(_this$styleRef$parent = this.styleRef.parentNode) === null || _this$styleRef$parent === void 0 || _this$styleRef$parent.removeChild(this.styleRef);
		}
		this.cleanEvents();
		this.props.zoomWithScroll && this.clearScrollEvent();
	}
	componentDidUpdate(prevProps) {
		var _prevProps$cropSize, _this$props$cropSize, _prevProps$cropSize2, _this$props$cropSize2, _prevProps$crop, _this$props$crop, _prevProps$crop2, _this$props$crop2;
		if (prevProps.rotation !== this.props.rotation) {
			this.computeSizes();
			this.recomputeCropPosition();
		} else if (prevProps.aspect !== this.props.aspect) this.computeSizes();
		else if (prevProps.objectFit !== this.props.objectFit) this.computeSizes();
		else if (prevProps.zoom !== this.props.zoom) this.recomputeCropPosition();
		else if (((_prevProps$cropSize = prevProps.cropSize) === null || _prevProps$cropSize === void 0 ? void 0 : _prevProps$cropSize.height) !== ((_this$props$cropSize = this.props.cropSize) === null || _this$props$cropSize === void 0 ? void 0 : _this$props$cropSize.height) || ((_prevProps$cropSize2 = prevProps.cropSize) === null || _prevProps$cropSize2 === void 0 ? void 0 : _prevProps$cropSize2.width) !== ((_this$props$cropSize2 = this.props.cropSize) === null || _this$props$cropSize2 === void 0 ? void 0 : _this$props$cropSize2.width)) this.computeSizes();
		else if (((_prevProps$crop = prevProps.crop) === null || _prevProps$crop === void 0 ? void 0 : _prevProps$crop.x) !== ((_this$props$crop = this.props.crop) === null || _this$props$crop === void 0 ? void 0 : _this$props$crop.x) || ((_prevProps$crop2 = prevProps.crop) === null || _prevProps$crop2 === void 0 ? void 0 : _prevProps$crop2.y) !== ((_this$props$crop2 = this.props.crop) === null || _this$props$crop2 === void 0 ? void 0 : _this$props$crop2.y)) this.emitCropAreaChange();
		if (prevProps.zoomWithScroll !== this.props.zoomWithScroll && this.containerRef) this.props.zoomWithScroll ? this.containerRef.addEventListener("wheel", this.onWheel, { passive: false }) : this.clearScrollEvent();
		if (prevProps.video !== this.props.video) {
			var _this$videoRef$curren3;
			(_this$videoRef$curren3 = this.videoRef.current) === null || _this$videoRef$curren3 === void 0 || _this$videoRef$curren3.load();
		}
		const objectFit = this.getObjectFit();
		if (objectFit !== this.state.mediaObjectFit) this.setState({ mediaObjectFit: objectFit }, this.computeSizes);
	}
	getAspect() {
		const { cropSize, aspect } = this.props;
		if (cropSize) return cropSize.width / cropSize.height;
		return aspect;
	}
	getObjectFit() {
		if (this.props.objectFit === "cover") {
			if ((this.imageRef.current || this.videoRef.current) && this.containerRef) {
				var _this$imageRef$curren3, _this$videoRef$curren4, _this$imageRef$curren4, _this$videoRef$curren5;
				this.containerRect = this.containerRef.getBoundingClientRect();
				const containerAspect = this.containerRect.width / this.containerRect.height;
				return (((_this$imageRef$curren3 = this.imageRef.current) === null || _this$imageRef$curren3 === void 0 ? void 0 : _this$imageRef$curren3.naturalWidth) || ((_this$videoRef$curren4 = this.videoRef.current) === null || _this$videoRef$curren4 === void 0 ? void 0 : _this$videoRef$curren4.videoWidth) || 0) / (((_this$imageRef$curren4 = this.imageRef.current) === null || _this$imageRef$curren4 === void 0 ? void 0 : _this$imageRef$curren4.naturalHeight) || ((_this$videoRef$curren5 = this.videoRef.current) === null || _this$videoRef$curren5 === void 0 ? void 0 : _this$videoRef$curren5.videoHeight) || 0) < containerAspect ? "horizontal-cover" : "vertical-cover";
			}
			return "horizontal-cover";
		}
		return this.props.objectFit;
	}
	onPinchStart(e) {
		const pointA = Cropper.getTouchPoint(e.touches[0]);
		const pointB = Cropper.getTouchPoint(e.touches[1]);
		this.lastPinchDistance = getDistanceBetweenPoints(pointA, pointB);
		this.lastPinchRotation = getRotationBetweenPoints(pointA, pointB);
		this.onDragStart(getCenter(pointA, pointB));
	}
	onPinchMove(e) {
		if (!this.currentDoc || !this.currentWindow) return;
		const pointA = Cropper.getTouchPoint(e.touches[0]);
		const pointB = Cropper.getTouchPoint(e.touches[1]);
		const center = getCenter(pointA, pointB);
		this.onDrag(center);
		if (this.rafPinchTimeout) this.currentWindow.cancelAnimationFrame(this.rafPinchTimeout);
		this.rafPinchTimeout = this.currentWindow.requestAnimationFrame(() => {
			const distance = getDistanceBetweenPoints(pointA, pointB);
			const newZoom = this.props.zoom * (distance / this.lastPinchDistance);
			this.setNewZoom(newZoom, center, { shouldUpdatePosition: false });
			this.lastPinchDistance = distance;
			const rotation = getRotationBetweenPoints(pointA, pointB);
			const newRotation = this.props.rotation + (rotation - this.lastPinchRotation);
			this.props.onRotationChange && this.props.onRotationChange(newRotation);
			this.lastPinchRotation = rotation;
		});
	}
	render() {
		var _this$state$mediaObje;
		const { image, video, mediaProps, cropperProps, transform, crop: { x, y }, rotation, zoom, cropShape, showGrid, roundCropAreaPixels, style: { containerStyle, cropAreaStyle, mediaStyle }, classes: { containerClassName, cropAreaClassName, mediaClassName } } = this.props;
		const objectFit = (_this$state$mediaObje = this.state.mediaObjectFit) !== null && _this$state$mediaObje !== void 0 ? _this$state$mediaObje : this.getObjectFit();
		return /* @__PURE__ */ react.createElement("div", {
			onMouseDown: this.onMouseDown,
			onTouchStart: this.onTouchStart,
			ref: (el) => this.containerRef = el,
			"data-testid": "container",
			style: containerStyle,
			className: classNames("reactEasyCrop_Container", containerClassName)
		}, image ? /* @__PURE__ */ react.createElement("img", _objectSpread2(_objectSpread2({
			alt: "",
			className: classNames("reactEasyCrop_Image", objectFit === "contain" && "reactEasyCrop_Contain", objectFit === "horizontal-cover" && "reactEasyCrop_Cover_Horizontal", objectFit === "vertical-cover" && "reactEasyCrop_Cover_Vertical", mediaClassName)
		}, mediaProps), {}, {
			src: image,
			ref: this.imageRef,
			style: _objectSpread2(_objectSpread2({}, mediaStyle), {}, { transform: transform || `translate(${x}px, ${y}px) rotate(${rotation}deg) scale(${zoom})` }),
			onLoad: this.onMediaLoad
		})) : video && /* @__PURE__ */ react.createElement("video", _objectSpread2(_objectSpread2({
			autoPlay: true,
			playsInline: true,
			loop: true,
			muted: true,
			className: classNames("reactEasyCrop_Video", objectFit === "contain" && "reactEasyCrop_Contain", objectFit === "horizontal-cover" && "reactEasyCrop_Cover_Horizontal", objectFit === "vertical-cover" && "reactEasyCrop_Cover_Vertical", mediaClassName)
		}, mediaProps), {}, {
			ref: this.videoRef,
			onLoadedMetadata: this.onMediaLoad,
			style: _objectSpread2(_objectSpread2({}, mediaStyle), {}, { transform: transform || `translate(${x}px, ${y}px) rotate(${rotation}deg) scale(${zoom})` }),
			controls: false
		}), (Array.isArray(video) ? video : [{ src: video }]).map((item) => /* @__PURE__ */ react.createElement("source", _objectSpread2({ key: item.src }, item)))), this.state.cropSize && /* @__PURE__ */ react.createElement("div", _objectSpread2({
			ref: this.cropperRef,
			style: _objectSpread2(_objectSpread2({}, cropAreaStyle), {}, {
				width: roundCropAreaPixels ? Math.round(this.state.cropSize.width) : this.state.cropSize.width,
				height: roundCropAreaPixels ? Math.round(this.state.cropSize.height) : this.state.cropSize.height
			}),
			tabIndex: 0,
			onKeyDown: this.onKeyDown,
			onKeyUp: this.onKeyUp,
			"data-testid": "cropper",
			className: classNames("reactEasyCrop_CropArea", cropShape === "round" && "reactEasyCrop_CropAreaRound", showGrid && "reactEasyCrop_CropAreaGrid", cropAreaClassName)
		}, cropperProps)));
	}
};
Cropper.defaultProps = {
	zoom: 1,
	rotation: 0,
	aspect: 4 / 3,
	maxZoom: MAX_ZOOM,
	minZoom: MIN_ZOOM,
	cropShape: "rect",
	objectFit: "contain",
	showGrid: true,
	style: {},
	classes: {},
	mediaProps: {},
	cropperProps: {},
	zoomSpeed: 1,
	restrictPosition: true,
	zoomWithScroll: true,
	keyboardStep: KEYBOARD_STEP
};
Cropper.getMousePoint = (e) => ({
	x: Number(e.clientX),
	y: Number(e.clientY)
});
Cropper.getTouchPoint = (touch) => ({
	x: Number(touch.clientX),
	y: Number(touch.clientY)
});
//#endregion
//#region src/index.ts
var src_default = Cropper;
//#endregion
exports.default = src_default;
exports.getInitialCropFromCroppedAreaPercentages = getInitialCropFromCroppedAreaPercentages;
exports.getInitialCropFromCroppedAreaPixels = getInitialCropFromCroppedAreaPixels;
